app/database/build_store.py

```diff
@@ -0,0 +1,307 @@
+    """
+    _logger = get_logger(__name__)
+    _cached_builds_by_id = OrderedDict()
+
+    @classmethod
+    def get(cls, build_id: int, allow_incompleted_builds=False) -> Optional[Build]:
+        """
+        Returns a build by id.
+        :param build_id: The id for the build whose status we are getting
+        """
+        build = cls._cached_builds_by_id.get(build_id)
+        if build is None:
+            cls._logger.info('Requested build (id: {}) was not found in cache, checking database.'.format(build_id))
+            build = cls._reconstruct_build(build_id, allow_incompleted_builds=allow_incompleted_builds)
+            if build is not None:
+                cls._cached_builds_by_id[build_id] = build
+                cls._logger.notice('Build (id: {}) was added to cache.'.format(build_id))
+
+        return build
+
+    @classmethod
+    def get_range(cls, start: int, end: int, allow_incompleted_builds=False) -> List[Build]:
+        """
+        Returns a list of all builds.
+        :param start: The starting index of the requested build.
+        :param end: 1 + the index of the last requested element, although if this is greater than the total number
+                    of builds available the length of the returned list may be smaller than (end - start).
+        """
+        builds = []
+        # Add 1 to start & end because we're create build_id's, not indices
+        for build_id in range(start + 1, end + 1):
+            try:
+                builds.append(cls.get(build_id, allow_incompleted_builds=allow_incompleted_builds))
+            except IncompleteBuild:
+                pass
+        return builds
+
+    @classmethod
+    def add(cls, build: Build):
+        """
+        Add new build to collection.
+        :param build: The build to add to the store.
+        """
+        build_id = cls._store_build(build)
+        build._build_id = build_id
+        cls._cached_builds_by_id[build_id] = build
+
+    @classmethod
+    def save(cls, build: Build):
+        """
+        Save current state of given build.
+        We assume that this build already exists in the database. You should always call `add` before
+        trying to ever call a `save`. This should already be done for you in cluster_master.
+        We also assume this build is already in the cache in its current state.
+        :param build: The build to save to database.
+        """
+        cls._logger.notice('Saving build (id: {}) in database.'.format(build.build_id()))
+        build.save()
+
+    @classmethod
+    def clean_up(cls):
+        """
+        Save current state of all cached builds.
+        """
+        session = Connection.get()
+        cls._logger.notice('Saving all active builds to database...')
+        for build_id in cls._cached_builds_by_id:
+            build = cls._cached_builds_by_id[build_id]
+            build.save()
+        session.commit()
+        cls._logger.notice('...done')
```

Comment:
These logging messages were more for my sake of making sure everything is working okay -- more than happy to clean them up